### ISSUE TITLE
EmptyLineSeparator ignores configured tokens: reports violations for unlisted tokens and skips listed ones

### ISSUE THREAD

**Issue by Tieshiyang** [checkstyle/checkstyle#17568]
I have read check documentation: https://checkstyle.sourceforge.io/checks/whitespace/emptylineseparator.html#EmptyLineSeparator
I have downloaded the latest checkstyle from: https://checkstyle.org/cmdline.html#Download_and_Run
I have executed the cli and showed it below, as cli describes the problem better than 1,000 words

Code:
```bash
/*
 * Copyright 2025
 */


package benchimarkfilesmakeup;
import java.util.List; // Expected violation
public class Test {
}
```
Config:
```
/var/tmp $ cat config.xml
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
          "-//Checkstyle//DTD Checkstyle Configuration 1.3//EN"
          "https://checkstyle.org/dtds/configuration_1_3.dtd">
<module name="Checker">
    <property name="localeLanguage" value="en"/>
    <!-- Checks for whitespace                               -->
    <!-- See http://checkstyle.org/checks/whitespace/index.html -->
    <module name="TreeWalker">
        <module name="EmptyLineSeparator">
            <property name="tokens" value="IMPORT"/>
            <property name="allowMultipleEmptyLines" value="false"/>
        </module>
    </module>
</module>
```
Cli:
```
/var/tmp $ java -jar checkstyle-10.26.1-all.jar -c config.xml Test.java
Starting audit...
[ERROR] /var/tmp/Test.java:8:1: 'CLASS_DEF' should be separated from previous line. [EmptyLineSeparator]
Audit done.
```

---

The [Checkstyle documentation](https://checkstyle.sourceforge.io/checks/whitespace/emptylineseparator.html#EmptyLineSeparator) states that:

“Checks for empty line separators before package, all import declarations, fields, constructors, methods, nested classes, static initializers and instance initializers.”

This suggests that only the tokens listed in the tokens property should be checked.

However, the behavior is inconsistent:
Observed issues:

False Positive (FP): It reports a violation before the CLASS_DEF line — even though CLASS_DEF is not in the token list.

False Negative (FN): It fails to report a violation for the IMPORT statement that is not preceded by an empty line — even though IMPORT is in the token list.
---

**Comment by romani** [comment]
it might demand  empty line between import and class, be cause import is your target.

whole logic is here https://github.com/checkstyle/checkstyle/blob/master/src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/EmptyLineSeparatorCheck.java

**Comment by VeniceWish** [comment]
Sorry for the late reply.
>“Checks for empty line separators before package, all import declarations, fields, constructors, methods, nested classes, static initializers and instance initializers.”

The doc says it checks **before**. When I set `tokens=IMPORT`, I expected a violation before IMPORT, but it didn’t report there and instead reported before `CLASS_DEF`.
This feels inconsistent between the doc and the actual behavior.

https://checkstyle.sourceforge.io/checks/whitespace/emptylineseparator.html#Example4-code shows that it works for methods and variables to demand separation Before.

**Comment by romani** [comment]
Definitely looks weird either doc or implementation bug should be fixed

**Comment by VeniceWish** [comment]
Thanks for confirming！

**Comment by Aman-Baliyan** [comment]
i am on it

**Comment by Aman-Baliyan** [comment]
@romani  please check this,

for Package_def token->

Example used-->

```
/*
 * Copyright 2025
 */
package temp;
import java.util.List;
import java.util.Set;

public class Test {

}
```

config ->

```
\src\temp>cat config.xml 
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
        "-//Checkstyle//DTD Checkstyle Configuration 1.3//EN"
        "https://checkstyle.org/dtds/configuration_1_3.dtd">

<module name="Checker">
    <property name="localeLanguage" value="en"/>

    <module name="TreeWalker">
        <module name="EmptyLineSeparator">
            <property name="tokens" value="PACKAGE_DEF"/>
            <property name="allowMultipleEmptyLines" value="false"/>
        </module>
    </module>
</module>
```

output -> C:\Users\amanc\Downloads\Ai_Interviewer\PracticeCheckStyle\src\temp>java -jar checkstyle-12.3.0-all.jar -c config.xml Test.java 

```
Starting audit...
[ERROR] C:\Users\amanc\Downloads\Ai_Interviewer\PracticeCheckStyle\src\temp\Test.java:4:1: 'package' should be separated from previous line. [EmptyLineSeparator]
[ERROR] C:\Users\amanc\Downloads\Ai_Interviewer\PracticeCheckStyle\src\temp\Test.java:5:1: 'import' should be separated from previous line. [EmptyLineSeparator]
Audit done.
Checkstyle ends with 2 errors.
```

-> Package-def check for empty line before and after the package.

For Import token->

Example used->

Test.java
```
/*
 * Copyright 2025
 */
package temp;
import java.util.List;
import java.util.Set;
public class Test {

}
```

config->
```
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
        "-//Checkstyle//DTD Checkstyle Configuration 1.3//EN"
        "https://checkstyle.org/dtds/configuration_1_3.dtd">

<module name="Checker">
    <property name="localeLanguage" value="en"/>

    <module name="TreeWalker">
        <module name="EmptyLineSeparator">
            <property name="tokens" value="IMPORT"/>
            <property name="allowMultipleEmptyLines" value="false"/>
        </module>
    </module>
</module>
```

output ->

```
C:\Users\amanc\Downloads\Ai_Interviewer\PracticeCheckStyle\src\temp>java -jar checkstyle-12.3.0-all.jar -c config.xml Test.java
Starting audit...
[ERROR] C:\Users\amanc\Downloads\Ai_Interviewer\PracticeCheckStyle\src\temp\Test.java:7:1: 'CLASS_DEF' should be separated from previous line. [EmptyLineSeparator]
Audit done.
Checkstyle ends with 1 errors.
```

-> means for import token it does not check space separator before it checks for space separator after it.

for class-def ->

example ->

```
/*
 * Copyright 2025
 */
package temp;
// comments
import java.util.List;
import java.util.Set;
public class Test {

}
class B{

}
```

config.xml ->

```
C:\Users\amanc\Downloads\Ai_Interviewer\PracticeCheckStyle\src\temp>cat config.xml
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
        "-//Checkstyle//DTD Checkstyle Configuration 1.3//EN"
        "https://checkstyle.org/dtds/configuration_1_3.dtd">

<module name="Checker">
    <property name="localeLanguage" value="en"/>

    <module name="TreeWalker">
        <module name="EmptyLineSeparator">
            <property name="tokens" value="CLASS_DEF"/>
            <property name="allowMultipleEmptyLines" value="false"/>
        </module>
    </module>
</module>
```

output ->

```
C:\Users\amanc\Downloads\Ai_Interviewer\PracticeCheckStyle\src\temp>java -jar checkstyle-12.3.0-all.jar -c config.xml Test.java
Starting audit...
[ERROR] C:\Users\amanc\Downloads\Ai_Interviewer\PracticeCheckStyle\src\temp\Test.java:11:1: 'CLASS_DEF' should be separated from previous line. [EmptyLineSeparator]
Audit done.
Checkstyle ends with 1 errors.
```


--> Class def is also checks for space after it

**Comment by Aman-Baliyan** [comment]
i think there is only problem with the documentation that the checks is not explained clearly because there is only package def which checks for separator before and after it all other tokens checks for separator after them.
Please tell me about your thoughts on this

**Comment by romani** [comment]
@Aman-Baliyan thanks a lot for very clear details on issue. Please send PR to explain logic. Please consider to update examples to show it by examples.

**Comment by romani** [comment]
fix is merged
